ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/TimeServiceBackend.scala

```diff
@@ -38,19 +45,46 @@ object TimeServiceBackend {
     }
   }
 
-  private class ObservingTimeServiceBackend(
-      timeProvider: TimeServiceBackend,
-      onTimeChange: Instant => Future[Unit]
-  ) extends TimeServiceBackend {
-    override def getCurrentTime: Instant = timeProvider.getCurrentTime
-
-    override def setCurrentTime(expectedTime: Instant, newTime: Instant): Future[Boolean] =
-      timeProvider
-        .setCurrentTime(expectedTime, newTime)
-        .flatMap { success =>
-          if (success)
-            onTimeChange(expectedTime).map(_ => true)(DirectExecutionContext)
-          else Future.successful(false)
-        }(DirectExecutionContext)
+  private final class AkkaQueueBasedObservedTimeServiceBackend(delegate: TimeServiceBackend)(
+      implicit materializer: Materializer
+  ) extends ObservedTimeServiceBackend {
+    private val queues = mutable.Set[SourceQueueWithComplete[Instant]]()
```

Comment:
What is the synchronization/locking mechanism that protects `queues` from inconsistent updates?

Reply:
Good question. Perhaps we should use a `concurrent.TrieMap` instead.

Reply:
Done.